ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs

```diff
@@ -16,27 +26,142 @@ import           Data.Maybe (isJust)
+  SomeSecond (BlockQuery blockQueryA) == SomeSecond (BlockQuery blockQueryB)
+    = SomeSecond blockQueryA == SomeSecond blockQueryB
+
+instance Show (SomeSecond BlockQuery blk) => Show (SomeSecond Query blk) where
+  show (SomeSecond (BlockQuery blockQueryA)) = "Query " ++ show (SomeSecond blockQueryA)
+
+queryEncodeNodeToClient ::
+    forall blk. (
+      SerialiseNodeToClient blk (SomeSecond BlockQuery blk)
+    )
+  => CodecConfig blk
+  -> NodeToClientVersion
+  -> BlockNodeToClientVersion blk
+  -> (SomeSecond Query blk)
+  -> Encoding
+queryEncodeNodeToClient codecConfig version blockVersion (SomeSecond query)
+  | not (version >= NodeToClientV_9)
+  = error "encode failure: Serializing Query is not supported"
+  | otherwise
+  = case query of
+    BlockQuery blockQuery ->
+      encodeTag 0 <> encodeNodeToClient
+                      @blk
+                      @(SomeSecond BlockQuery blk)
+                      codecConfig
+                      blockVersion
+                      (SomeSecond blockQuery)
+
+queryDecodeNodeToClient ::
+    forall blk. (
+      SerialiseNodeToClient blk (SomeSecond BlockQuery blk)
+    )
+  => CodecConfig blk
+  -> NodeToClientVersion
+  -> BlockNodeToClientVersion blk
+  -> forall s. Decoder s (SomeSecond Query blk)
+queryDecodeNodeToClient codecConfig version blockVersion
+  | not (version >= NodeToClientV_9)
+  = fail $ "decode failure: Deserializing Query is not supported (on node to client version: " ++ show version ++ ")"
+  | otherwise
+  = do
+    tag <- decodeTag
+    case tag of
+      0 -> do
+        SomeSecond x <- decodeNodeToClient
+            @blk
+            @(SomeSecond BlockQuery blk)
+            codecConfig
+            blockVersion
+        return (SomeSecond (BlockQuery x))
+      _ -> fail $ "SomeSecond Query blk: unknown tag " ++ show tag
+
+queryEncodeResult ::
+     forall blk result. SerialiseResult blk (BlockQuery blk)
+  => CodecConfig blk
+  -> NodeToClientVersion
+  -> BlockNodeToClientVersion blk
+  -> Query blk result
+  -> result
+  -> Encoding
+queryEncodeResult codecConfig version blockNodeToClientVersion query result
+  | not (version >= NodeToClientV_9)
+  = error $ "encode failure: Serializing Query result is not supported (on node to client version: " ++ show version ++ ")"
+  | otherwise
+  = case query of
+      BlockQuery blockQuery ->
+        encodeTag 0
+        <> encodeResult codecConfig blockNodeToClientVersion blockQuery result
+
+queryDecodeResult ::
+     forall blk result. SerialiseResult blk (BlockQuery blk)
+  => CodecConfig blk
+  -> NodeToClientVersion
+  -> BlockNodeToClientVersion blk
+  -> Query blk result
+  -> forall s. Decoder s result
+queryDecodeResult codecConfig version blockNodeToClientVersion query
+  | not (version >= NodeToClientV_9)
+  = error $ "encode failure: Deserializing Query result is not supported (on node to client version: " ++ show version ++ ")"
```

Comment:
Same as my comment on `queryEncodeResult`; however, we will need backward compatibility, in general, not just `error` when the negotiated version is less than `V_9` (or whatever the threshold ends up being). We will still need to be able to communicate with "older" nodes.